ion-c-sys/tests/value.rs

```diff
@@ -57,116 +57,157 @@ use Val::*;
 #[derive(Debug)]
 struct TestCase {
     lit: &'static str,
+    bin: &'static [u8],
     elem: Elem,
 }
```

Comment:
It feels like `rstest` is exerting some weird pressure here. The `TestCase` struct has enough information to run a full test case, but you can't actually do anything with it until `rstest` gives you more information in the form of a `TestMode`. That seems backwards from how I would probably set it up--but maybe that's just the easiest way to "go with the grain" in `rstest`. This is where I appreciate using a `@DataProvider` (TestNG) or `@Parameterized` (JUnit) for data-driven tests: so that I can explode the tests according to my own logic, rather than the framework's.

If I were starting from scratch, your `TestCase` seems like a good place to begin, but would probably change `elem` to `expected` and `input` to be a vec containing 1 or more representations that should yield that `expected` value once they're consumed by the reader. Usually that means 1 of each, but are there cases where you could end up with the same `expected` from more than one text or binary input value? Are there `expected` values that can be gotten form a binary input, but not a text one? I think that's where the tests start to get really interesting.

Reply:
The thinking of the test case design (given the compile-time limitation of rstest)  was to have a 3-way equivalence between a sort of DOM-like representation, text, and binary, and use that as a template to drive reading and a writing test matrix with maybe some flavor of rountrip testing in the mix.

Reply:
The real problem is that Rust's test framework is purely static--you can't dynamically create tests like various test frameworks can, so you see a cottage industry of `proc_macro` to simulate it at compile time.  You can get some of the way there, but it is weak.

We *could* do a binary crate with our own test driver, but tooling integration would be weak (debugging a single test case, IDE integration, etc.).  I don't know if this route is worth it, but it is an option.